dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/InactivityTimeoutHandler.java

```diff
@@ -23,27 +23,55 @@
  * {@link IdleStateHandler}. 
  */
 public class InactivityTimeoutHandler extends IdleStateHandler {
-    
+
     private static final TraceComponent tc = Tr.register(InactivityTimeoutHandler.class, TCPMessageConstants.NETTY_TRACE_NAME, 
             TCPMessageConstants.TCP_BUNDLE);
 
+    private boolean waitingForInitialRead = true;
+
     public InactivityTimeoutHandler(long readerIdleTime, long writerIdleTime, long allIdleTime, TimeUnit unit) {
         super(readerIdleTime, writerIdleTime, allIdleTime, unit);
     }
 
     @Override
     protected void channelIdle(ChannelHandlerContext ctx, IdleStateEvent evt) throws Exception{
-        // Overriden to call the userEventTriggered of this handler instead of the next to keep the timeout logic contained to one class/handler
-        userEventTriggered(ctx, evt);
+        // Overridden to call the userEventTriggered of this handler instead of the next to keep the timeout logic contained to one class/handler.
+
+        // If we are waitingForInitialRead then we can timeout twice! This mimics the Channel Framework double read on initial read behavior.
+        if(waitingForInitialRead && !evt.isFirst()) { // On the first idleStateEvent if we have not already read then continue waiting.
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(tc, "calling userEventTriggered(1)");
+            }
+            userEventTriggered(ctx, evt);
+        } else if(!waitingForInitialRead) { // If we've timed out and already performed a read then we are in fact in a timeout scenario
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(tc, "calling userEventTriggered(2)");
+            }
+            userEventTriggered(ctx, evt);
+        }
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        // Are we the initial read? If so then we want to extend the timeout times 2 to mimic the behavior on Channel Framework which retries on time on the initial read if there is an
+        // initial timeout.
+        if(waitingForInitialRead) {
+            waitingForInitialRead = false;
+        }
+
+        ctx.fireChannelRead(msg);
```

Comment:
Look at other areas where `channelRead` is overriden and see if I'm missing anything.